src/ArrayInterface.jl

```diff
@@ -755,8 +762,41 @@ include("static.jl")
 include("ranges.jl")
 include("indexing.jl")
 include("dimensions.jl")
+include("axes.jl")
+include("size.jl")
 include("stridelayout.jl")
 
+
+abstract type AbstractArray2{T,N} <: AbstractArray{T,N} end
+
+Base.size(A::AbstractArray2) = ArrayInterface.size(A)
+Base.size(A::AbstractArray2, dim) = ArrayInterface.size(A, dim)
+
+Base.axes(A::AbstractArray2) = ArrayInterface.axes(A)
+Base.axes(A::AbstractArray2, dim) = ArrayInterface.axes(A, dim)
+
+Base.strides(A::AbstractArray2) = ArrayInterface.strides(A)
+Base.strides(A::AbstractArray2, dim) = ArrayInterface.strides(A, dim)
```

Comment:
Should the size and stride methods convert to `Int` to avoid potential type instabilities/problems in generic code unaware of ArrayInterface?

Reply:
These only exist to pass things forward to the ArrayInterface implementation. But if you look at the actually implementation...

```julia
axes(a, dim) = axes(a, to_dims(a, dim))
function axes(a::A, dim::Integer) where {A}
    if parent_type(A) <: A
        return Base.axes(a, Int(dim))
    else
        return axes(parent(a), to_parent_dims(A, dim))
    end
end
```

...`StaticInt` is never passed to a method in Base but persists internally, preventing premature conversion of `StaticInt` to an `Int` and relying on constant propagation for inference.

Reply:
I'd I define a statically sized `AbstractArray2` using the ArtayInterface, then `Base.size` will suddenly return `StaticInt`s as well.

Reply:
I thought it made sense because `StaticInt` is part of the public interface and we want to propagate static information, avoiding a dependence on constant propagation when dealing with deeply nested types. I'm sure we will run into some situations where methods explicitly require `Int`, but I also assumed that the end goal was for `StaticInt` to become standard.

Reply:
My thinking was that users might assume that `size` returns an `NTuple{N,Int}`, and end up inadvertently writing code like `myprod` that doesn't handle heterogenous tuples well:
```julia
julia> using ArrayInterface: StaticInt

julia> sz = (StaticInt(4), 5);

julia> @code_warntype prod(sz)
MethodInstance for prod(::Tuple{StaticInt{4}, Int64})
  from prod(x::Tuple{Any, Vararg{Any}}) in Base at tuple.jl:480
Arguments
  #self#::Core.Const(prod)
  x::Tuple{StaticInt{4}, Int64}
Body::Int64
1 ─ %1 = Core._apply_iterate(Base.iterate, Base.:*, x)::Int64
└──      return %1


julia> function myprod(x)
           p = 1
           for a ∈ x
               p *= a
           end
           p
       end
myprod (generic function with 1 method)

julia> @code_warntype myprod(sz)
MethodInstance for myprod(::Tuple{StaticInt{4}, Int64})
  from myprod(x) in Main at REPL[10]:1
Arguments
  #self#::Core.Const(myprod)
  x::Tuple{StaticInt{4}, Int64}
Locals
  @_3::Union{Nothing, Tuple{Union{StaticInt{4}, Int64}, Int64}}
  p::Int64
  a::Union{StaticInt{4}, Int64}
Body::Int64
1 ─       (p = 1)
│   %2  = x::Tuple{StaticInt{4}, Int64}
│         (@_3 = Base.iterate(%2))
│   %4  = (@_3::Core.Const((static(4), 2)) === nothing)::Core.Const(false)
│   %5  = Base.not_int(%4)::Core.Const(true)
└──       goto #4 if not %5
2 ┄ %7  = @_3::Union{Tuple{StaticInt{4}, Int64}, Tuple{Int64, Int64}}
│         (a = Core.getfield(%7, 1))
│   %9  = Core.getfield(%7, 2)::Int64
│         (p = p * a)
│         (@_3 = Base.iterate(%2, %9))
│   %12 = (@_3 === nothing)::Bool
│   %13 = Base.not_int(%12)::Bool
└──       goto #4 if not %13
3 ─       goto #2
4 ┄       return p


julia> @btime myprod($(Ref(sz))[])
  14.331 ns (1 allocation: 16 bytes)
20

julia> @btime prod($(Ref(sz))[])
  0.989 ns (0 allocations: 0 bytes)
20
```
I don't want to accidentally cause bad performance in other libraries. I have no idea how prevalent code like this is though, and would hope authors would fix it.

I also figure that if code is set up to take advantage of static size information, it could probably also take the step to use `ArrayInterface.size` & co instead. E.g., many of my libraries have `using ArrayInterface: size, strides`.

But if code is written in ignorance of these, is it more likely that they'd benefit from the forced constant prop, or that they'd be hurt by the need for constant prop to ensure type stability?

As is, I have definitions like [here](https://github.com/chriselrod/StrideArraysCore.jl/blob/master/src/ptr_array.jl#L115):
```julia
@inline Base.size(A::AbstractStrideArray) = map(Int, size(A))
@inline Base.strides(A::AbstractStrideArray) = map(Int, strides(A))

@inline create_axis(s, ::Zero) = CloseOpen(s)
@inline create_axis(s, ::One) = One():s
@inline create_axis(s, o) = CloseOpen(s, s+o)

@inline ArrayInterface.axes(A::AbstractStrideArray) = map(create_axis, size(A), offsets(A))
@inline Base.axes(A::AbstractStrideArray) = axes(A)

@inline ArrayInterface.offsets(A::PtrArray) = getfield(getfield(A, :ptr), :offsets)
@inline ArrayInterface.static_length(A::AbstractStrideArray) = prod(size(A))

# type stable, because index known at compile time
@inline type_stable_select(t::NTuple, ::StaticInt{N}) where {N} = t[N]
@inline type_stable_select(t::Tuple, ::StaticInt{N}) where {N} = t[N]
# type stable, because tuple is homogenous
@inline type_stable_select(t::NTuple, i::Integer) = t[i]
# make the tuple homogenous before indexing
@inline type_stable_select(t::Tuple, i::Integer) = map(Int, t)[i]

@inline function ArrayInterface.axes(A::AbstractStrideVector, i::Integer)
    if i == 1
        o = type_stable_select(offsets(A), i)
        s = type_stable_select(size(A), i)
        return create_axis(s, o)
    else
        return One():1
    end
end
@inline function ArrayInterface.axes(A::AbstractStrideArray, i::Integer)
    o = type_stable_select(offsets(A), i)
    s = type_stable_select(size(A), i)
    create_axis(s, o)
end
@inline Base.axes(A::AbstractStrideArray, i::Integer) = axes(A, i)
```
Meant to try and avoid placing too many performance foot-guns about.

Reply:
Thanks for the examples Chris. I think it nicely illustrates the problems with returning `StaticInt` in the wild.

It will undoubtedly cause problems introducing `StaticInt` into the public space, although StaticArrays already does this and returns `SOneTo` for its axes. Perhaps another way of approaching this is asking ourselves if we really do plan for something like `StaticInt` to be passed around in the future. If we do then at what point do we start pushing people to make that adjustment? Having to move back and forth between static and dynamic space so that users don't have to learn about using things like `NTuple{N}` vs `Tuple{Vararg{Any,N}}` seems like a mistake to me, but so does country music and I'm not going to take that away from people either.

@timholy, you had some valuable input when we made `StaticInt`. Any opinions?

Reply:
I made it so that `strides` and `size` will return `Int` types. Returning statically sized axes seems to work for StaticArrays so I left that in. I think this also makes it so that new subtypes of `AbstractArray2` can just define `axes` and `axes_types` and `ArrayInterface.size` and `ArrayInterface.strides` can catch any static info.